waku/v2/waku_discv5.nim

```diff
@@ -48,6 +48,36 @@ type WakuDiscoveryV5* = ref object
     protocol*: protocol.Protocol
     listening*: bool
 
+func topicsToRelayShards*(topics: seq[string]): Result[Option[RelayShards], string] =
```

Comment:
perhaps its me but i find this code a bit confusing. there are some special cases i == 0, i > 0, and not sure i like the use of prevTopic.


im for sure missing something so take this as just **pseudocode**. we iterate multiple times but performance here is not an issue, but imho it improves readibility?

```nim
# check errors
if parsedTopicsRes.anyIt(it.isErr()):
    return err(it.error)

# check all are named sharding
if parsedTopicsRes.anyIt(it.get.kind == NsPubsubTopicKind.StaticSharding):
    return error(xxx)

# check all shards belong to the same cluster
if !parsedTopicsRes.allIt(it.get == parsedTopicsRes[0].get.cluster):
   return errors(xxx)

# return
return RelayShards.init(
                 cluster=parsedTopicsRes[0].get.cluster,
                 shards=parsedTopicsRes.mapIt(it.get.shard)

```

Reply:
Agree with Alvaro here re readability. And I think it's possible to do only one iteration if you expand your predicate, e.g.:

```
if not parsedTopicsRes.allIt(
  it.isOk() &&
  it.get.kind == NsPubsubTopicKind.StaticSharding &&
  it.get.cluster == parsedTopicsRes[0].get.cluster
  )
  return error(....)
```

If you want to you could even extract the three conditions above into their own predicate function (e.g. `isValid()`), but I think it's readable enough without having to do this. Up to you :)

Reply:
```
if parsedTopicsRes.anyIt(it.isErr()):
    return err(it.error)
```
That is what I wanted to do but seams like you can't use ```it``` outside of scope no?

Reply:
Where do you mean out of scope? The predicate function could be defined inside the scope of the outer `topicsToRelayShards()` func.

The following should work structurally:

```
proc myOuterFunction()...
  
  func isValidPredicate(someVar: VariableType): bool =
    return someBooleanPredicateCondition

  if not myList.allIt(it.isValidPredicate()):
   return error(...)
```

Not great pseudocode, but hope it illustrates what I mean 😅

Of course, in this instance I don't even think the separate predicate `func` is necessary and you could just add the simple conditions within the `allIt()` argument directly.

Reply:
In Rust I would do
```
let result: Result<Vec<_>, Error> = topics.into_iter().map(|item| parse(item)).collect;
let parsed_topics = result?;
```
Can't seams to do the same in Nim... How do you return an Error inside a sequence in a functional style?
```
for res in parsedTopics
   if res.IsErr()
      return err(res.error)
```
That works but it's imperative style and looks ugly :P